002 tidyr <-> stringr alignment (#15)
tidyverse/tidyups: 002-tidyr-stringr.Rmd

---

# Tidyup 2: tidyr \<-> stringr alignment

**Champion**: Hadley Wickham  
**Co-champion**: Mine Çetinkaya-Rundel  
**Status**: Draft

## Abstract

tidyr provides a handful functions like `separate()` and `extract()` that perform string manipulation.
Unfortunately, these functions are only weakly connected to their stringr equivalents, making them harder to learn and remember.
This tidyup proposes a new cohesive family of string manipulation functions to tackle common problems that arise during data tidying.
The existing functions will remain so existing code is unaffected, but will be superseded to steer folks towards a more consistent API.

```{r setup}
library(tidyr)
library(stringr)
```

## Motivation

stringr and tidyr provide surprisingly different interfaces for string manipulation.
This means that common tasks require integrating two distinct syntaxes and learning more about stringr doesn't improve your use of tidyr.

The root cause of the problem is that stringr was designed before I understood the importance of data frames, so many stringr functions return matrices (or lists of matrices), which makes them hard to use with other tidyverse packages.

> No stringr function returns a data frame, which is particularly painful given that `summarise()` and `mutate()` can now use data frame results to add multiple columns.

Reviewer: Is the issue that stringr functions don't return data frames? Or that they don't return vectors? I might be thinking about this wrong but forcats functions also don't return data frames, for example, but they're a lot easier to use in pipelines than string functions.

Author: I think it's something like some stringr functions need to return multiple "things", and in that case it's now obvious that you should be returning a data frame instead of a matrix.

To overcome this weakness, tidyr provides a few functions that work directly with data frames.
These take a data frame and column specification, and return a modified data frame, typically with new columns specified by the call.
Unfortunately, the tidyr functions have little relationship to their stringr equivalents:

-   `extract()` is similar to `str_match()`
-   `separate()` with regex is similar to `str_split_fixed()`

> -   `separate()` with numeric columns has no equivalent in stringr.

Reviewer: ```suggestion
-   `separate()` with numeric columns has no equivalent in stringr
```

-   `separate_rows()` is equivalent to `str_split()`

> Additionally, tidyr uses base R regular expressions, not stringr.

Reviewer: Is it "base R regular expressions" or "base R string manipulation functions"?

Reviewer: Also, how about giving an example of a tidyr function as an "e.g." here that doesn't use stringr under the hood? Something like "`tidyr::___()` uses `___()` instead of `stringr::___()`"? (Looking at the code for some of the tidyr functions, many seem to define a `str_*()` function and use it, so I wasn't sure what would be a good example here.)

Author: I mean the "base R regexp expression engine" as opposed to the "ICU regexp engine" (as used by stringi).  It's definitely hard to follow the logic but e.g. `separate()` calls `str_separate()` which calls `str_split_fixed()` which calls `str_split_n()` which calls `gregexpr()`.

> There are relatively few difference between regexp engines, but you need a very different syntax to control options (e.g. ignoring case).

Reviewer: ```suggestion
There are relatively few differences between the regexp engines, but you need very different syntax to control options (e.g. ignoring case).
```

> While it's important to keep tidyr dependencies lightweight (and installing stringr can be a challenge in some environments), it would be nice to use it if available, following the footsteps laid by [Radix Ordering in](https://github.com/tidyverse/tidyups/blob/main/003-dplyr-radix-ordering.md) [`dplyr::arrange()`](https://github.com/tidyverse/tidyups/blob/main/003-dplyr-radix-ordering.md).

Reviewer: "can be a challenge in some environments due to __?__" (I'm not sure what goes in the blank but would be nice to say that here, if possible to state concisely.)

Additionally there are a few gaps in the current API:

-   Neither stringr nor tidyr provide a particularly nice way to work with named capture groups.
    There's also scope for more directly matching column names to patterns if `extract()` didn't attempt to match the syntax of `separate()` so closely, e.g. `str_extract(x, c(type = ".*", "|", fruit = "orange|apple"))`.

-   tidyr only provides one function for splitting a string across rows, rather than columns: `separate_rows()`.

>     What other functions should exist in this family?

Reviewer: When you say "in this family" do you mean "other functions for splitting a string across rows"? I'm not sure what is meant by "family", i.e. which rows of the table in the subsequent "Components" section the family is referring to.

Author: Yeah, here I mean family = "split string across rows" (as opposed to columns)

-   Into a varying number of pieces by length, e.g. `"abcdef"` -> `c("ab", "cd", "ef")`

-   Into a fixed number of pieces by a pattern, e.g. `"a/b/c"` -> `c(treatment = "a", group = "b", subgroup = "c").`

-   Into a varying number of pieces by a pattern, e.g. `"a/b/c"` -> `c("a", "b", "c").`

-   Into named groups, e.g. `"a(b)c"` -> `c(treatment = "a", group = "b", subgroup = "c")`

Matching these operations up with the existing stringr and tidyr function gives the following table:

| Split by | Length  | stringr         | tidyr             |
|----------|---------|-----------------|-------------------|
| Length   | Fixed   |                 | `separate()`      |
| Length   | Varying |                 |                   |
| Pattern  | Fixed   | `str_split_n()` | `separate()`      |
| Pattern  | Varying | `str_split()`   | `separate_rows()` |
| Group    | Fixed   | `str_match()`   | `extract()`       |

Note that `separate()` is repeated twice, and there are a number of empty cells.
Interestingly, there's no stringr function that allows you to extract multiple pieces by position; you currently have to call `str_sub()` multiple times yourself.

> Generally, when the pieces might vary from entry to entry you'll want to create rows; when the pieces are fixed, you'll want to create columns.

Reviewer: ```suggestion
Generally, when the numbers of pieces might vary from entry to entry you'll want to create rows; when the numbers of pieces are fixed, you'll want to create columns.
```

This might not be true in every case, but I think it's true in enough cases that we can assume it as a constraint for this problem.

> (Please let me know if you have a counter-example.)

Reviewer: I agree that when the numbers of pieces might vary from entry to entry it's reasonable to assume that you'll want to create rows. I'm not sure I as strongly agree that when they're fixed you'll want to create columns.

Suppose I have the following data frame:

```r
id value   
 1 "10,20,15"
 2 "20,10,12"
 3 "18,15,11"
```

I  think I'm equally likely to want

```r
id value_1 value_2 value_3
 1 10      20      15     
 2 20      10      12     
 3 18      15      11  
```

as I am to want

```r
id value value_no
 1 10           1
 1 20           2
 1 15           3
 2 20           1
 2 10           2
 2 12           3
 3 18           1
 3 15           2
 3 11           3
```

as a result.

This makes me think I should have the option to use something like `separate_wider()` or `separate_longer()` as opposed to only `separate_auto()` (à la `unnest_auto()`) that will make the decision for me.

Author: It's something like: this case is just a special case of the "varying number of pieces" problem where the pieces all have the same number.

Author: Putting fixed number of pieces into rows is easy; I'm more concerned about the case where you might want to put varying number of pieces in to columns.

Reviewer: Right, I imagine you'll need to define a default value like `NA`?

> We don't need to consider compound problems that need to be solved by splitting into rows and columns simultaneously because you a compound problem (e.g. `"a-b-c/d-e-f/g-h-i"`) can always be solved it by first splitting into rows and then into columns.

Reviewer: ```suggestion
We don't need to consider compound problems that need to be solved by splitting into rows and columns simultaneously because a compound problem (e.g. `"a-b-c/d-e-f/g-h-i"`) can always be solved by first splitting into rows and then into columns.
```

---
output: github_document
---

# Tidyup 2: tidyr \<-> stringr alignment

**Champion**: Hadley Wickham  
**Co-champion**: Mine Çetinkaya-Rundel  
**Status**: Draft

```{r, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

## Abstract

> tidyr provides a handful functions like `separate()` and `extract()` that perform string manipulation.

Reviewer 2: ```suggestion
tidyr provides a handful of functions like `separate()` and `extract()` that perform string manipulation.
```

```{r setup}
library(tidyr)
library(stringr)
```

## Motivation

stringr and tidyr provide surprisingly different interfaces for string manipulation.
This means that common tasks require integrating two distinct syntaxes and learning more about stringr doesn't improve your use of tidyr.

The root cause of the problem is that stringr was designed before I understood the importance of data frames, so stringr functions that need to return multiple results use matrices rather than data frames.
This is particularly painful since dplyr 1.0.0 gave `summarise()` and `mutate()` the ability to create multiple columns from data frame results.

To overcome this weakness, tidyr provides a few functions that work directly with data frames.
These take a data frame and column specification, and return a modified data frame, typically with new columns specified by the call.
Unfortunately, the tidyr functions have little relationship to their stringr equivalents:

-   `extract()` is similar to `str_match()`.
-   `separate()` with regex is similar to `str_split_fixed()`.
-   `separate()` with numeric `col` (which splits by position) has no equivalent in stringr.

> -   `separate_rows()` is equivalent to `str_split()`

Reviewer 2: ```suggestion
-   `separate_rows()` is equivalent to `str_split()`.
```

Additionally there are a few gaps in the current API:

-   Neither stringr nor tidyr provide a particularly nice way to work with named capture groups.
    There's also scope for more directly matching column names to patterns if `extract()` didn't attempt to match the syntax of `separate()` so closely, e.g. `str_extract(x, c(type = ".*", "|", fruit = "orange|apple"))`.

-   tidyr only provides one function for splitting a string into rows, not columns: `separate_rows()`.
    What other tidyr functions should create rows?

-   `tidyr::separate()` can split up a string by position or by pattern.
    This now seems too clever, and makes it hard to notice that there are two different tools here.

## Solutions

Finding a solution to this problem first requires a careful analysis of the cases.
I think we can break splitting down into two largely independent components:

-   How do you specify the pieces?

    -   By length/position, e.g. `c(1, 5, 10)`.
    -   By a regular expression pattern, e.g. `","`, `"\s+"`.
    -   With regular expression "groups" formed by `()`,`"a(b)c".`

> -   How do the piece vary from element to element?

Reviewer 3: ```suggestion
-   How do the pieces vary from element to element?
```

    -   Does each string have the same number of components which we might name and put in columns?

>     -   Does each string potential have a varying number of components which we'd put into rows?

Reviewer 2: ```suggestion
    -   Does each string potentially have a varying number of components which we'd put into rows?
```

    `_at` family was deprecated in dplyr so some possibility for confusion; but it's probably fairly clear that the context is different here.

    I don't love `group`, because I don't think many people think about regexp `()` as creating groups (even though that's the technical name), but there doesn't seem to be any obvious better word.
    This function also feels a little different to "splitting" which is what lead to `tidyr::extract()` but I think it's better to keep the root verb the same since there are otherwise so many similarities.

-   A second suffix should distinguish between splitting into fixed/named and varying/unnamed.
    If we assume a strong connection between fixed/named and columns and varying/unnamed and rows, the obvious suffixes are `wider` and `longer`.

    The return type associated with these suffixes will be a little different between tidyr and stringr.
    For stringr, `wider` implies a data frame and `longer` implies a list of character vectors.
    For tidyr, `wider` implies growing the columns and `longer` implies growing the rows.

>     This will also influence the function arguments since `wider` functions names.

Reviewer 2: unfinished thought?

```{r}
library(dplyr, warn.conflicts = FALSE)

df <- tibble(x = c("a-b", "c-d"))
df %>% separate(x, c("first", "second"), "-")

df %>% 
  mutate(
    str_separate_by_wider(x, "-", c("first", "second")), 
    .keep = "unused"
  )
```

This works because `mutate()` will create multiple columns if an unnamed calculation returns a data frame.

However, there's no straightforward way to use `str_separate_by_longer()` to create multiple columns.
The simplest pure dplyr approach is probably to use `rowwise():`

```{r}
df %>%
  rowwise() %>% 
  summarise(str_separate_by_longer(x, "-")[[1]])
```

But it's quite hard to explain why this works, and the use of `summarise()` means that you need to explicitly list any variables that you want to leave as is.
This possibly speaks to a missing verb in dplyr (e.g. <https://github.com/tidyverse/dplyr/issues/5874>) but it's also possible that creating rows is fundamentally a tidyr operation.
Either way, there's no easy fix in sight, and it seems unappealing to force folks to learn sophisticated dplyr idioms in order to tackle a routine data tidying task.

### Arguments

-   `str_separate_*_*()`: first argument is character vector.

> -   `str_*_*_wider()` needs a `col_names` argument and arguments that controls what happens if there are too many/too few pieces.

Reviewer 2: ```suggestion
-   `str_*_*_wider()` needs a `col_names` argument and arguments that control what happens if there are too many/too few pieces.
```

    Probably needs `names_sep` for consistency with `unnest_wider()`.

```{=html}
<!-- -->
```
-   `str_separate_group_wider()` has a few options for specifying names:

    -   Named groups `"(?<type>.*)|(?<fruit>orange|apple)`

    -   Named vector: `c(type = ".*", "|", fruit = "orange|apple"))`

-   Should `convert` continue to exist?
    Should it use `type.convert()` or `readr::type_convert()`?

>     How to override imputed types?

Reviewer 2: I much prefer the finer control offered by `tidyr::pivot_longer(values_transform = )`. Could be `into_transform =`.

Author: Or just `transform` to match `unnest_wider()`?

The vast majority of the work will occur in the stringr functions, and the tidyr functions will inherit their arguments.
`separate_*_*()` will work similarly by instead of taking character vector, takes pair of data frame + tidyselect specification.

> Also has `remove` argument which controls whether or not the input column is removed.

Reviewer: When you say "Also has" do you mean "All of these functions will also have..."?

There are relatively few differences between the engines, but the syntax for controlling regular expression options (e.g. ignoring case) is very different.
In the past, we have kept stringr out of tidyr dependencies because it can be expensive/painful to build stringi on linux.
However, most tidyverse users will have stringr installed, and RSPM makes stringi easier to install, so it would be nice to use it if available, following the footsteps laid by [Radix Ordering in](https://github.com/tidyverse/tidyups/blob/main/003-dplyr-radix-ordering.md) [`dplyr::arrange()`](https://github.com/tidyverse/tidyups/blob/main/003-dplyr-radix-ordering.md).

Additionally there are a few gaps in the current API:

-   Neither stringr nor tidyr provide a particularly nice way to work with named capture groups.
    There's also scope for more directly matching column names to patterns if `extract()` didn't attempt to match the syntax of `separate()` so closely, e.g. `str_extract(x, c(type = ".*", "|", fruit = "orange|apple"))`.

-   tidyr only provides one function for splitting a string into rows, not columns: `separate_rows()`.

>     What other tidyr functions should create rows?

Reviewer 2: Ok I see you have the `Eliminate tidyr functions` section below, but maybe we can think about it one more time with `unchop()`

>     What other tidyr functions should create rows?

Reviewer: I think the well-named tidyr helpers are useful, even if the same task can be accomplished with dplyr+stringr, particularly for common tasks, and I think the examples here are common data wrangling tasks. Some of the dplyr+stringr combinations can be difficult to learn early on because you have to think both across columns and down the rows at the same time to understand what the code is doing and what results to expect.

>     What other tidyr functions should create rows?

Reviewer 2: I think I am mainly excited about the fact that we can accomplish this with dplyr + `unchop()` / `unpack()` at all.

AFAIK, we previously didn't think we had an existing solution for what `str_*_longer()` + dplyr + `unchop()` would do - so now that we do have a way to perform that operation it is at least clear what the tidyr implementation of things like `separate_by_longer()` will be. I agree that this combination could be difficult for new R users to come up with on their own, so it is probably fine for the tidyr variants to exist.

I was a little excited about the idea that tidyr could get out of the string manipulation business entirely, since `separate()` and `extract()` never felt quite right in tidyr, but that is probably overambitious 😬

>     What other tidyr functions should create rows?

Author: Yeah, I think the main take-away is that the tidyr functions will be mostly thin wrappers around the stringr functions which do the hard work of implementation.

Overall, this proposal only reduces the total number of functions by two, in return for a small gain.

### Eliminate tidyr functions

Can we eliminate the tidyr wrappers altogether and just use dplyr + stringr?
This seemed like the "obvious" solution to me at least twice, so I wanted to make sure to explain why it doesn't work.

To make the discussion concrete I'll start with quick and dirty implementations of `str_separate_by_wider()` and `str_separate_by_longer()`:

```{r}
str_separate_by_wider <- function(x, pattern, into) {
  mat <- str_split_fixed(x, pattern, length(into))
  colnames(mat) <- into
  tibble::as_tibble(mat)
}
str_separate_by_longer <- function(x, pattern) {
  str_split(x, pattern)
}

str_separate_by_wider(c("a-b", "c-d"), "-", c("first", "second"))
str_separate_by_longer(c("a-b", "c-d-e"), "-")
```

We can use `str_separate_by_wider()` directly with `mutate()` to achieve the same affect as `separate()`:

```{r}
library(dplyr, warn.conflicts = FALSE)

df <- tibble(x = c("a-b", "c-d"))
df %>% separate(x, c("first", "second"), "-")

df %>% 
  mutate(
    str_separate_by_wider(x, "-", c("first", "second")), 
    .keep = "unused"

>   )

Reviewer 2: And here is my multi-column approach with `str_separate_by_wider()`

``` r
library(tidyverse)
library(stringr)

df <- tibble(
  x = 1:2,
  y = c("a-c", "c-d"),
  z = c("q-r", "z-f")
)

str_separate_by_wider <- function(x, pattern, into) {
  mat <- str_split_fixed(x, pattern, length(into))
  colnames(mat) <- into
  tibble::as_tibble(mat)
}

df %>%
  mutate(across(c(y, z), ~str_separate_by_wider(.x, "-", c("first", "second")))) %>%
  unpack(c(y, z), names_sep = "_")
#> # A tibble: 2 × 5
#>       x y_first y_second z_first z_second
#>   <int> <chr>   <chr>    <chr>   <chr>   
#> 1     1 a       c        q       r       
#> 2     2 c       d        z       f
```

<sup>Created on 2021-12-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

```

This works because `mutate()` will create multiple columns if an unnamed calculation returns a data frame.

However, there's no straightforward way to use `str_separate_by_longer()` to create multiple columns.
The simplest pure dplyr approach is probably to use `rowwise():`

```{r}
df %>%
  rowwise() %>% 
  summarise(str_separate_by_longer(x, "-")[[1]])
```

But it's quite hard to explain why this works, and the use of `summarise()` means that you need to explicitly list any variables that you want to leave as is.

> This possibly speaks to a [missing verb](https://github.com/tidyverse/dplyr/issues/5874) in dplyr but it's also possible that creating rows is fundamentally a tidyr operation.

Reviewer 2: > it's also possible that creating rows is fundamentally a tidyr operation.

I feel like this is correct, sort of. I think it has to do with the fact that tidyr operations are good at taking a single column and either:
- Widening it to make more columns
- Lengthening it to make more rows (and this is what is useful here)

Either way, there's no easy fix in sight, and it seems unappealing to force folks to learn sophisticated dplyr idioms in order to tackle a routine data tidying task.

> 

Reviewer 2: I think this is the biggest reason we'd push forward with the tidyr variants. How complex is `across()` + either `unpack()` or `unchop()`?

Could `unnest()` be used instead in most cases? Since it is a hybrid of unpack and unchop and more people are familiar with it.

Reviewer: I think `across()` + `unpack()` or `unchop()` is quite complex, especially if the alternative is a single verb that accomplishes a common task.
